rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,430 @@
+    SSHLogin
+    // User is a teleport username
+    User string `json:"user"`
+    // RequestID is a uuid for the request
+    RequestID string
+    // Command is the client command being requested
+    Command string
+}
+
+// SSHLogin contains common SSH login parameters.
+type SSHLogin struct {
+    // ProxyAddr is the target proxy address
+    ProxyAddr string
+    // PubKey is SSH public key to sign
+    PubKey []byte
+    ...
+}
+
+// SSHLoginResponse is a user login response
+type SSHLoginResponse struct {
+    // Username contains the username for the login certificates
+    Username string `json:"username"`
+    // Cert is a PEM encoded SSH certificate signed by SSH certificate authority
+    Cert []byte `json:"cert"`
+    // TLSCertPEM is a PEM encoded TLS certificate signed by TLS certificate authority
+    TLSCert []byte `json:"tls_cert"`
+    // HostSigners is a list of signing host public keys trusted by proxy
+    HostSigners []TrustedCerts `json:"host_signers"`
+}
+```
+
+#### `rpc GetHeadlessLoginRequest`
+
+This endpoint is used by Teleport clients to retrieve headless login request details before prompting the user for approval/denial.
+
+The endpoint is only authorized for the user who requested Headless authentication (and for server roles).
+
+```proto
+service AuthService {
+  rpc GetHeadlessLoginRequest(GetHeadlessLoginRequestRequest) returns (HeadlessLoginRequest);
+}
+
+message GetHeadlessLoginRequestRequest {
+  // RequestID is the headless login request uuid
+  string RequestID = 1;
+}
+
+message HeadlessLoginRequest {
+  // ID is the headless login request uuid
+  string ID = 1;
+  // User is a teleport user name
+  string User = 2;
+  // PubKey is SSH public key to sign
+  bytes PubKey = 3;
+  // Command is a `tsh` command in plain text
+  string Command = 4;
+  // State is the headless login request state
+  State State = 5;
+}
+
+// State is a headless login request state
+enum State {
+  PENDING = 0;
+  DENIED = 1;
+  APPOVED = 2;
+}
+```
+
+#### `rpc UpdateHeadlessLoginRequestState`
+
+This endpoint is used by Teleport clients to update headless login request state to approved or denied. If the client requests to approve, the client will receive an MFA challenge and will need to reply with a valid MFA challenge response. To support this multi-step request, the rpc provides an rpc stream to both sides of the connection.
+
+The endpoint is only authorized for the user who requested Headless authentication (and for server roles).
+
+```proto
+service AuthService {
+  rpc UpdateHeadlessLoginRequestState(stream UpdateHeadlessLoginRequestStateRequest) returns (stream UpdateHeadlessLoginRequestStateResponse);
+}
+
+message UpdateHeadlessLoginRequestStateRequest {
+  oneof Request {
+    // Init is the initial request
+    UpdateHeadlessLoginRequestStateRequestInit Init = 1;
+    // MFAResponse is the client's signed MFA challenge response
+    MFAAuthenticateResponse MFAResponse = 2;
```

Comment:
Just a note on a future implementation: 
the code should verify that it is not possible to send `MFAAuthenticateResponse` without calling first `UpdateHeadlessLoginRequestStateRequestInit`